x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/lib/search_fields.tsx

```diff
@@ -0,0 +1,227 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import React from 'react';
+
+import { NormalizedFields, NormalizedField, SearchResult, SearchMetadata } from '../types';
+import { ALL_DATA_TYPES } from '../constants';
+
+interface FieldWithMeta {
+  field: NormalizedField;
+  metadata: SearchMetadata;
+}
+
+interface SearchData {
+  term: string;
+  searchRegexArray: RegExp[];
+  type?: string;
+}
+
+interface FieldData {
+  path: string;
+  type: string;
+}
+
+const sortResult = (a: FieldWithMeta, b: FieldWithMeta) => {
+  if (a.metadata.score > b.metadata.score) {
+    return -1;
+  } else if (b.metadata.score > a.metadata.score) {
+    return 1;
+  }
+  if (a.metadata.stringMatch === null) {
+    return 1;
+  } else if (b.metadata.stringMatch === null) {
+    return -1;
+  }
+  // With the same score, the longest match string wins.
+  return b.metadata.stringMatch.length - a.metadata.stringMatch.length;
+};
+
+const calculateScore = (metadata: Omit<SearchMetadata, 'score' | 'display'>): number => {
+  let score = 0;
+
+  if (metadata.matchPath) {
+    score += 15;
+  }
+
+  if (metadata.matchStartOfPath) {
+    score += 5;
+  }
+
+  if (metadata.fullyMatchPath) {
+    score += 5;
+  }
+
+  if (metadata.matchType) {
+    score += 5;
+  }
+
+  if (metadata.fullyMatchType) {
+    score += 5;
+  }
+
+  return score;
+};
+
+const getJSXdisplayFromMeta = (
+  searchData: SearchData,
+  fieldData: FieldData,
+  metadata: Omit<SearchMetadata, 'score' | 'display'>
+): JSX.Element => {
+  const { term } = searchData;
+  const { path } = fieldData;
+
+  let display: JSX.Element = <span>{path}</span>;
+
+  if (metadata.fullyMatchPath) {
+    display = (
+      <span>
+        <strong>{path}</strong>
+      </span>
+    );
+  } else if (metadata.matchStartOfPath) {
+    const endString = path.substr(term.length, path.length);
+    display = (
+      <span>
+        <strong>{term}</strong>
+        {endString}
+      </span>
+    );
+  } else if (metadata.matchPath) {
+    const { stringMatch } = metadata;
+    const charIndex = path.toLowerCase().indexOf(stringMatch!.toLowerCase());
+    const startString = path.substr(0, charIndex);
+    const endString = path.substr(charIndex + stringMatch!.length);
+    display = (
+      <span>
+        {startString}
+        <strong>{stringMatch}</strong>
+        {endString}
+      </span>
+    );
+  }
+
+  return display;
+};
+
+const getSearchMetadata = (searchData: SearchData, fieldData: FieldData): SearchMetadata => {
+  const { term, type, searchRegexArray } = searchData;
+  const typeToCompare = type ?? term;
+
+  const matchStartOfPath = fieldData.path.startsWith(term);
+  const fullyMatchPath = term === fieldData.path;
+  const matchType = fieldData.type.includes(typeToCompare);
+  const fullyMatchType = typeToCompare === fieldData.type;
```

Comment:
Not sure I follow, it would be replacing this:

```js
const matchStartOfPath = fieldData.path.startsWith(term);
```

with this

```js
const indexMatch = fieldData.path.indexOf(term);
const matchStartOfPath = indexMatch === 0;
```

And then repeat for `fieldData.type`. Not sure if there is any gain (apart adding lines). For the "whether the term matches the entire string" I just compare both strings. Am I missing something?

Reply:
So it's this line:
```ts
term === fieldData.path
```

That I think potentially does more work than if we used `String.prototype.indexOf` and just checked for `indexMatch === 0`.

So not massive gains, but it's more this that I am thinking of: https://stackoverflow.com/a/26554820

Reply:
Cool, thanks. I will look into it 👍